<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Collection of utility functions for the pairwise dotplots">

<title>dotplot_util – comandos</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ed96de9b727972fe78a7b5d16c58bf87.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-000f0825b9c55ed21d14970d810c14a9.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="dotplot_util – comandos">
<meta property="og:description" content="Collection of utility functions for the pairwise dotplots">
<meta property="og:site_name" content="comandos">
<meta name="twitter:title" content="dotplot_util – comandos">
<meta name="twitter:description" content="Collection of utility functions for the pairwise dotplots">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">comandos</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./dotplot_util.html">dotplot_util</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ComAnDOS</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./util.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">util</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">genes</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">plot</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dotplot_util.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">dotplot_util</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./report.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">report</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#highlight_cluster" id="toc-highlight_cluster" class="nav-link active" data-scroll-target="#highlight_cluster">highlight_cluster</a></li>
  <li><a href="#unique_genes" id="toc-unique_genes" class="nav-link" data-scroll-target="#unique_genes">unique_genes</a></li>
  <li><a href="#map_to_colormap" id="toc-map_to_colormap" class="nav-link" data-scroll-target="#map_to_colormap">map_to_colormap</a></li>
  <li><a href="#map_array_to_color" id="toc-map_array_to_color" class="nav-link" data-scroll-target="#map_array_to_color">map_array_to_color</a></li>
  <li><a href="#add_homology_context" id="toc-add_homology_context" class="nav-link" data-scroll-target="#add_homology_context">add_homology_context</a></li>
  <li><a href="#plot_dotplot" id="toc-plot_dotplot" class="nav-link" data-scroll-target="#plot_dotplot">plot_dotplot</a></li>
  <li><a href="#add_connections" id="toc-add_connections" class="nav-link" data-scroll-target="#add_connections">add_connections</a></li>
  <li><a href="#make_dotplot" id="toc-make_dotplot" class="nav-link" data-scroll-target="#make_dotplot">make_dotplot</a></li>
  <li><a href="#plot_colorbar_legend" id="toc-plot_colorbar_legend" class="nav-link" data-scroll-target="#plot_colorbar_legend">plot_colorbar_legend</a></li>
  <li><a href="#plot_dot_legend" id="toc-plot_dot_legend" class="nav-link" data-scroll-target="#plot_dot_legend">plot_dot_legend</a></li>
  <li><a href="#get_dot_color" id="toc-get_dot_color" class="nav-link" data-scroll-target="#get_dot_color">get_dot_color</a></li>
  <li><a href="#get_dot_size" id="toc-get_dot_size" class="nav-link" data-scroll-target="#get_dot_size">get_dot_size</a></li>
  <li><a href="#feature_colors" id="toc-feature_colors" class="nav-link" data-scroll-target="#feature_colors">feature_colors</a></li>
  <li><a href="#gene_order" id="toc-gene_order" class="nav-link" data-scroll-target="#gene_order">gene_order</a></li>
  <li><a href="#calculate_adjacency_matrix" id="toc-calculate_adjacency_matrix" class="nav-link" data-scroll-target="#calculate_adjacency_matrix">calculate_adjacency_matrix</a></li>
  <li><a href="#label_pos" id="toc-label_pos" class="nav-link" data-scroll-target="#label_pos">label_pos</a></li>
  <li><a href="#prepare_dotplot" id="toc-prepare_dotplot" class="nav-link" data-scroll-target="#prepare_dotplot">prepare_dotplot</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/galicae/comandos/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">dotplot_util</h1>
</div>

<div>
  <div class="description">
    Collection of utility functions for the pairwise dotplots
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>This module contains a collection of utility functions for the pairwise dotplots, as well as some more general functions that are used in the dotplot generation process. I wrote documentation for the functions here, but I didn’t write tests for them. I’m not sure this page is very valuable to end users.</p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L121" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="highlight_cluster" class="level3">
<h3 class="anchored" data-anchor-id="highlight_cluster">highlight_cluster</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> highlight_cluster(</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    clusters:ndarray, <span class="co"># The array of cluster names.</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    cluster:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># The cluster to highlight (default: None).</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    bg:Union<span class="op">=</span><span class="st">'black'</span>, <span class="co"># The background color. Accepts all matplotlib-compatible color formats (default: "black").</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    hl:Union<span class="op">=</span><span class="st">'red'</span>, <span class="co"># The highlight color. Accepts all matplotlib-compatible color formats (default: "red").</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>ndarray: <span class="co"># The array of colors, with the same length as `clusters`.</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Highlight a cluster in a list of clusters by setting the color of the cluster to <code>hl</code> and the color of the rest to <code>bg</code>.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L89" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="unique_genes" class="level3">
<h3 class="anchored" data-anchor-id="unique_genes">unique_genes</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> unique_genes(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    connections:ndarray, <span class="co"># The array of connections. Columns should be (query_genes, target_genes, connection_strength). The last column is optional.</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>ndarray: <span class="co"># The array of unique genes.</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Extract the unique gene names from an array of connections.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L57" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="map_to_colormap" class="level3">
<h3 class="anchored" data-anchor-id="map_to_colormap">map_to_colormap</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> map_to_colormap(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    x:ndarray, <span class="co"># The array to map.</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    cmap:Union<span class="op">=</span><span class="st">'magma_r'</span>, <span class="co"># The color map to use. Should be a matplotlib colormap object or a string with the name of a matplotlib colormap (default: "magma_r").</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    vmin:<span class="bu">float</span><span class="op">=</span><span class="dv">0</span>, <span class="co"># The value to obtain the minimum color in the colormap. Should be &lt;= np.min(x) to avoid truncation (default: 0).</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    vmax:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># The value to obtain the maximum color in the colormap. Should be &gt;= np.max(x), and will use np.max(x) if set to `None` (default: None).</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>ndarray: <span class="co"># Array of RGBA values with a shape of ``x.shape + (4, )``.</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Map an array of values to a color palette.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L31" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="map_array_to_color" class="level3">
<h3 class="anchored" data-anchor-id="map_array_to_color">map_array_to_color</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> map_array_to_color(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    x:ndarray, <span class="co"># The array to map.</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    palette:Colormap, <span class="co"># The color map to use. Should be a matplotlib colormap object.</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    xmax:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># The maximum value to use for normalization. Should be &gt;= np.max(x), and will use np.max(x) if set to `None` (default: None).</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>ndarray: <span class="co"># Array of RGBA values with a shape of ``x.shape + (4, )``.</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Map an array of values to a color palette.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L1003" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_homology_context" class="level3">
<h3 class="anchored" data-anchor-id="add_homology_context">add_homology_context</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> add_homology_context(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    connections:ndarray, <span class="co"># The connections between genes. The columns should be (query_gene, target_gene).</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    orthology:DataFrame, <span class="co"># The orthology information as a DataFrame.</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>ndarray: <span class="co"># The connections array with homology context added. The columns will be (query_gene,</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>target_gene, connection_strength), <span class="kw">and</span> the values <span class="kw">in</span> connection_strength will depend on the</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>content of the orthology DataFrame.</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Add homology context to the given connections based on the orthology information.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L830" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="plot_dotplot" class="level3">
<h3 class="anchored" data-anchor-id="plot_dotplot">plot_dotplot</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_dotplot(</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    query_avg_expr:ndarray, <span class="co"># Array representing the average expression values of query genes.</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    target_avg_expr:ndarray, <span class="co"># Array representing the average expression values of target genes.</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    query_perc_expr:ndarray, <span class="co"># Array representing the percentage expression values of query genes.</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    target_perc_expr:ndarray, <span class="co"># Array representing the percentage expression values of target genes.</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    query_genes:List, <span class="co"># List of query gene names.</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    target_genes:List, <span class="co"># List of target gene names.</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    connections:ndarray, <span class="co"># An array where each row contains two genes and (optionally) the strength of their</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>connection.</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    query_cluster_colors:Dict, <span class="co"># Dictionary mapping query cluster names to their colors.</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    target_cluster_colors:Dict, <span class="co"># Dictionary mapping target cluster names to their colors.</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>    query_gene_colors:Dict, <span class="co"># Dictionary mapping query gene names to their colors.</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>    target_gene_colors:Dict, <span class="co"># Dictionary mapping target gene names to their colors.</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>    query_species:<span class="bu">str</span>, <span class="co"># Species name of the query genes.</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    target_species:<span class="bu">str</span>, <span class="co"># Species name of the target genes.</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    x_offset:<span class="bu">float</span><span class="op">=</span><span class="dv">1</span>, <span class="co"># Offset for the x-axis (default: 1).</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>    y_offset:<span class="bu">float</span><span class="op">=</span><span class="dv">0</span>, <span class="co"># Offset for the y-axis (default: 0).</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>    grid_offset:<span class="bu">int</span><span class="op">=</span><span class="dv">30</span>, <span class="co"># Offset for the grid spacing (default: 30).</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    query_clustering:<span class="bu">str</span><span class="op">=</span><span class="st">'leiden'</span>, <span class="co"># Clustering method for the query genes (default: "leiden").</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>    target_clustering:<span class="bu">str</span><span class="op">=</span><span class="st">'leiden'</span>, <span class="co"># Clustering method for the target genes (default: "leiden").</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>    output:<span class="bu">str</span><span class="op">=</span><span class="st">'./paired_dotplot.png'</span>, <span class="co"># Output file path for the plot (default: "./paired_dotplot.png").</span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>    title:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Title of the plot (default: None).</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>    title_font_size:<span class="bu">int</span><span class="op">=</span><span class="dv">16</span>, <span class="co"># Font size of the plot title (default: 16).</span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>    center:<span class="bu">bool</span><span class="op">=</span><span class="va">True</span>, <span class="co"># Whether to center the dotplots when the number of genes exceeds the maximum (default: True).</span></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>    cmap:Colormap<span class="op">=</span><span class="st">'magma_r'</span></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span><span class="va">None</span>:</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Plot the paired dotplot based on the given data.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L757" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_connections" class="level3">
<h3 class="anchored" data-anchor-id="add_connections">add_connections</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> add_connections(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    fig:Figure, <span class="co"># The paired dotplot figure to which the connections will be added.</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    connections:ndarray, <span class="co"># An array where each row contains two genes and (optionally) the strength of their</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>connection.</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    query_gene_names:List, <span class="co"># The list of query gene names.</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    query_gene_colors:Dict, <span class="co"># The dictionary mapping query gene names to their colors.</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    label_offset:<span class="bu">float</span>, <span class="co"># The offset for label positioning.</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span><span class="va">None</span>:</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Add connections between genes to the given paired dotplot figure.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L698" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="make_dotplot" class="level3">
<h3 class="anchored" data-anchor-id="make_dotplot">make_dotplot</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> make_dotplot(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    ax:Axes, <span class="co"># The Axes object on which to create the dotplot.</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    avg:ndarray, <span class="co"># The average expression values.</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    perc:ndarray, <span class="co"># The percentage expression values.</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    gene_names:List, <span class="co"># The list of gene names.</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    species:<span class="bu">str</span>, <span class="co"># The species name.</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    clustering:<span class="bu">str</span>, <span class="co"># The clustering information.</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    clust_color:List, <span class="co"># The list of colors for clusters.</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    gene_color:List, <span class="co"># The list of colors for genes.</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    side:<span class="bu">str</span><span class="op">=</span><span class="st">'left'</span>, <span class="co"># The side to place the y-axis labels, either "left" or "right" (default: "left").</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    cmap:Colormap<span class="op">=</span><span class="st">'magma_r'</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span><span class="va">None</span>:</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Make a dotplot on the given Axes object based on the average and percentage expression values.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L669" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="plot_colorbar_legend" class="level3">
<h3 class="anchored" data-anchor-id="plot_colorbar_legend">plot_colorbar_legend</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_colorbar_legend(</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    cbar_legend:Axes, <span class="co"># The Axes object representing the colorbar legend.</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    query_avg_expr:ndarray, <span class="co"># Array representing the average expression values of query genes.</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    target_avg_expr:ndarray, <span class="co"># Array representing the average expression values of target genes.</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    cmap:Colormap<span class="op">=</span><span class="st">'magma_r'</span>, <span class="co"># The Colormap instance or registered colormap name used to map scalar data to colors</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>(default: <span class="st">"magma_r"</span>).</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span><span class="va">None</span>:</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Plot the colorbar legend based on the average expression values of query and target genes.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L605" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="plot_dot_legend" class="level3">
<h3 class="anchored" data-anchor-id="plot_dot_legend">plot_dot_legend</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_dot_legend(</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    dot_legend, <span class="co"># The subplot of the grid that contains the dotplot legend.</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    size_exponent:<span class="bu">float</span><span class="op">=</span><span class="fl">1.5</span>, <span class="co"># The exponent to raise the fraction of cells in a group to, to get the dot size. The default</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="kw">is</span> <span class="fl">1.5</span>.</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    dot_size:<span class="bu">int</span><span class="op">=</span><span class="dv">200</span>, <span class="co"># The size of the largest dot. The default is 200.</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Create the dotplot legend, explaining dot size.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L542" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_dot_color" class="level3">
<h3 class="anchored" data-anchor-id="get_dot_color">get_dot_color</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_dot_color(</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    query:AnnData, <span class="co"># The query dataset.</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    target:AnnData, <span class="co"># The target dataset.</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    query_clustering:<span class="bu">str</span>, <span class="co"># The .obs column name to use for the query dataset.</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    target_clustering:<span class="bu">str</span>, <span class="co"># The .obs column name to use for the target dataset.</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    query_genes:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Array of query genes to subset the data, if any. If None, use all genes (default: None).</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    target_genes:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Array of target genes to subset the data, if any. If None, use all genes (default: None).</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    query_gene_names:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Array of query gene names (default: None).</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    target_gene_names:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Array of target gene names (default: None).</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    layer:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># The layer to use for the average expression calculation. If not specified, it will use the</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>`.X` slot of the `AnnData` objects. It <span class="kw">is</span> vital to <span class="bu">set</span> this correctly to avoid calculating</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>average expression on log1p<span class="op">-</span>transformed data (default: <span class="va">None</span>).</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>Tuple: <span class="co"># A tuple containing the dot color values for query and target datasets, respectively.</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Calculate average expression in each cluster and translate that to dot color for the dotplot.</em> Note that this function does not know what you did with the matrix before; if you have log-transformed the data it will calculate an average of logs, not the log of the exp-transformed average.</p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L475" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_dot_size" class="level3">
<h3 class="anchored" data-anchor-id="get_dot_size">get_dot_size</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_dot_size(</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    query:DataFrame, <span class="co"># The query dataset.</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    target:DataFrame, <span class="co"># The target dataset.</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    query_clustering:<span class="bu">str</span>, <span class="co"># The .obs column name to use for the query dataset.</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    target_clustering:<span class="bu">str</span>, <span class="co"># The .obs column name to use for the target dataset.</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    query_genes:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Array of query genes to subset the data, if any. If None, use all genes (default: None).</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    target_genes:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Array of target genes to subset the data, if any. If None, use all genes (default: None).</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    query_gene_names:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Array of query gene names (default: None).</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    target_gene_names:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Array of target gene names (default: None).</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>Tuple: <span class="co"># A tuple containing the dot size values for query and target datasets, respectively.</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Calculate which percentage of cells in each cluster express each gene, and translate that to dot</em> size for the dotplot.</p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L434" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="feature_colors" class="level3">
<h3 class="anchored" data-anchor-id="feature_colors">feature_colors</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> feature_colors(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    components:ndarray, <span class="co"># The array of components.</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    query_G:<span class="bu">int</span>, <span class="co"># The number of components for the query genes.</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    seed:<span class="bu">int</span><span class="op">=</span><span class="dv">42</span>, <span class="co"># The seed value for the random number generator (default: 42).</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>Tuple: <span class="co"># A tuple containing the colored components for query genes and target genes, respectively.</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Assign colors to the components based on the given array of components.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L393" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="gene_order" class="level3">
<h3 class="anchored" data-anchor-id="gene_order">gene_order</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> gene_order(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    full_adjacency:ndarray, <span class="co"># The full adjacency matrix represented as a 2D numpy array.</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    components:ndarray, <span class="co"># An array representing the components.</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    query_G:<span class="bu">int</span>, <span class="co"># The number of query genes.</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>Tuple: <span class="co"># A tuple containing the query gene order and the target gene order as numpy arrays.</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Calculate the order of genes based on the given full adjacency matrix and components. Highly</em> connected genes are placed first, genes without any connections are randomly ordered in the bottom of the plot.</p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L347" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="calculate_adjacency_matrix" class="level3">
<h3 class="anchored" data-anchor-id="calculate_adjacency_matrix">calculate_adjacency_matrix</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calculate_adjacency_matrix(</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    connections:ndarray, <span class="co"># The 2D array representing the connections between genes. Each row contains two gene</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>identifiers indicating a connection, <span class="kw">and</span> optionally the strength of that connection.</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    query_genes:List, <span class="co"># A list of genes that act as queries.</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    target_genes:List, <span class="co"># A list of genes that act as targets.</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>ndarray: <span class="co"># The adjacency matrix represented as a 2D numpy array. It has dimensions (query_G + target_G)</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>x (query_G <span class="op">+</span> target_G), where query_G <span class="kw">and</span> target_G are the lengths of query_genes <span class="kw">and</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>target_genes, respectively.</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Calculate the adjacency matrix based on the given connections, query genes, and target genes.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L315" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="label_pos" class="level3">
<h3 class="anchored" data-anchor-id="label_pos">label_pos</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> label_pos(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    display_coords:Dict, <span class="co"># A dictionary that holds the window extents of tick labels.</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    key:<span class="bu">str</span>, <span class="co"># The label to retrieve; a gene name.</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    side:<span class="bu">str</span><span class="op">=</span><span class="st">'left'</span>, <span class="co"># One of "left" or "right"; depending on orientation will return the leftmost or rightmost</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>position of the label (default: <span class="st">"left"</span>).</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>Tuple: <span class="co"># A tuple containing the x and y coordinates of the label.</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Get the edge coordinates of a label. Keep either the left or the right end of the word.</em></p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L266" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="prepare_dotplot" class="level3">
<h3 class="anchored" data-anchor-id="prepare_dotplot">prepare_dotplot</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> prepare_dotplot(</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    avg_expr:DataFrame, <span class="co"># Data frame that holds average expression for all genes and all clusters.</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    perc_expr:DataFrame, <span class="co"># Data frame that tracks the percentage of cells expressing each gene in every cluster.</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    cmap:Union<span class="op">=</span><span class="st">'magma_r'</span>, <span class="co"># The Colormap instance or registered colormap name used to map scalar data to colors</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>(default: <span class="st">"magma_r"</span>).</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    vmin:<span class="bu">float</span><span class="op">=</span><span class="dv">0</span>, <span class="co"># Minimum average expression value to show (default: 0).</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    vmax:Optional<span class="op">=</span><span class="va">None</span>, <span class="co"># Maximum average expression value to show (default: maximum average expr. value).</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    size_exponent:<span class="bu">float</span><span class="op">=</span><span class="fl">1.5</span>, <span class="co"># Dot size is computed as fraction ** size_exponent * dot_size (default: 1.5).</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    dot_size:<span class="bu">float</span><span class="op">=</span><span class="dv">200</span>, <span class="co"># The size of the largest dot (default: 200).</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>Tuple: <span class="co"># A tuple containing the melted average expression data frame, the melted percentage</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>expression data frame, <span class="kw">and</span> the array of RGBA<span class="op">-</span>coded color values <span class="cf">for</span> the average expression</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="kw">in</span> a cluster<span class="op">/</span>gene combination, according to the <span class="bu">input</span> color <span class="bu">map</span>.</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Pivots average expression and percent expressed tables to make them dotplot-friendly.</em></p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/galicae\.github\.io\/comandos");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/galicae/comandos/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>