<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Collection of utility functions for the pairwise dotplots">

<title>comandos - dotplot_util</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="comandos - dotplot_util">
<meta property="og:description" content="Collection of utility functions for the pairwise dotplots">
<meta property="og:site_name" content="comandos">
<meta name="twitter:title" content="comandos - dotplot_util">
<meta name="twitter:description" content="Collection of utility functions for the pairwise dotplots">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">comandos</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./dotplot_util.html">dotplot_util</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ComAnDOS</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./util.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">util</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">genes</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">plot</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dotplot_util.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">dotplot_util</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./report.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">report</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#highlight_cluster" id="toc-highlight_cluster" class="nav-link active" data-scroll-target="#highlight_cluster">highlight_cluster</a></li>
  <li><a href="#unique_genes" id="toc-unique_genes" class="nav-link" data-scroll-target="#unique_genes">unique_genes</a></li>
  <li><a href="#map_to_colormap" id="toc-map_to_colormap" class="nav-link" data-scroll-target="#map_to_colormap">map_to_colormap</a></li>
  <li><a href="#map_array_to_color" id="toc-map_array_to_color" class="nav-link" data-scroll-target="#map_array_to_color">map_array_to_color</a></li>
  <li><a href="#add_homology_context" id="toc-add_homology_context" class="nav-link" data-scroll-target="#add_homology_context">add_homology_context</a></li>
  <li><a href="#plot_dotplot" id="toc-plot_dotplot" class="nav-link" data-scroll-target="#plot_dotplot">plot_dotplot</a></li>
  <li><a href="#add_connections" id="toc-add_connections" class="nav-link" data-scroll-target="#add_connections">add_connections</a></li>
  <li><a href="#make_dotplot" id="toc-make_dotplot" class="nav-link" data-scroll-target="#make_dotplot">make_dotplot</a></li>
  <li><a href="#plot_colorbar_legend" id="toc-plot_colorbar_legend" class="nav-link" data-scroll-target="#plot_colorbar_legend">plot_colorbar_legend</a></li>
  <li><a href="#plot_dot_legend" id="toc-plot_dot_legend" class="nav-link" data-scroll-target="#plot_dot_legend">plot_dot_legend</a></li>
  <li><a href="#get_dot_color" id="toc-get_dot_color" class="nav-link" data-scroll-target="#get_dot_color">get_dot_color</a></li>
  <li><a href="#get_dot_size" id="toc-get_dot_size" class="nav-link" data-scroll-target="#get_dot_size">get_dot_size</a></li>
  <li><a href="#feature_colors" id="toc-feature_colors" class="nav-link" data-scroll-target="#feature_colors">feature_colors</a></li>
  <li><a href="#gene_order" id="toc-gene_order" class="nav-link" data-scroll-target="#gene_order">gene_order</a></li>
  <li><a href="#calculate_adjacency_matrix" id="toc-calculate_adjacency_matrix" class="nav-link" data-scroll-target="#calculate_adjacency_matrix">calculate_adjacency_matrix</a></li>
  <li><a href="#label_pos" id="toc-label_pos" class="nav-link" data-scroll-target="#label_pos">label_pos</a></li>
  <li><a href="#prepare_dotplot" id="toc-prepare_dotplot" class="nav-link" data-scroll-target="#prepare_dotplot">prepare_dotplot</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/galicae/comandos/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">dotplot_util</h1>
</div>

<div>
  <div class="description">
    Collection of utility functions for the pairwise dotplots
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>This module contains a collection of utility functions for the pairwise dotplots, as well as some more general functions that are used in the dotplot generation process. I wrote documentation for the functions here, but I didn’t write tests for them. I’m not sure this page is very valuable to end users.</p>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L124" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="highlight_cluster" class="level3">
<h3 class="anchored" data-anchor-id="highlight_cluster">highlight_cluster</h3>
<blockquote class="blockquote">
<pre><code> highlight_cluster (clusters:numpy.ndarray, cluster:Optional[str]=None,
                    bg:Union[str,tuple]='black',
                    hl:Union[str,tuple]='red')</code></pre>
</blockquote>
<p>Highlight a cluster in a list of clusters by setting the color of the cluster to <code>hl</code> and the color of the rest to <code>bg</code>.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>clusters</td>
<td>ndarray</td>
<td></td>
<td>The array of cluster names.</td>
</tr>
<tr class="even">
<td>cluster</td>
<td>typing.Optional[str]</td>
<td>None</td>
<td>The cluster to highlight (default: None).</td>
</tr>
<tr class="odd">
<td>bg</td>
<td>typing.Union[str, tuple]</td>
<td>black</td>
<td>The background color. Accepts all matplotlib-compatible color formats (default: “black”).</td>
</tr>
<tr class="even">
<td>hl</td>
<td>typing.Union[str, tuple]</td>
<td>red</td>
<td>The highlight color. Accepts all matplotlib-compatible color formats (default: “red”).</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td></td>
<td><strong>The array of colors, with the same length as <code>clusters</code>.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L92" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="unique_genes" class="level3">
<h3 class="anchored" data-anchor-id="unique_genes">unique_genes</h3>
<blockquote class="blockquote">
<pre><code> unique_genes (connections:numpy.ndarray)</code></pre>
</blockquote>
<p>Extract the unique gene names from an array of connections.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>connections</td>
<td>ndarray</td>
<td>The array of connections. Columns should be (query_genes, target_genes, connection_strength). The last column is optional.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td><strong>The array of unique genes.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L60" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="map_to_colormap" class="level3">
<h3 class="anchored" data-anchor-id="map_to_colormap">map_to_colormap</h3>
<blockquote class="blockquote">
<pre><code> map_to_colormap (x:numpy.ndarray,
                  cmap:Union[str,matplotlib.colors.Colormap]='magma_r',
                  vmin:float=0, vmax:Optional[float]=None)</code></pre>
</blockquote>
<p>Map an array of values to a color palette.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>x</td>
<td>ndarray</td>
<td></td>
<td>The array to map.</td>
</tr>
<tr class="even">
<td>cmap</td>
<td>typing.Union[str, matplotlib.colors.Colormap]</td>
<td>magma_r</td>
<td>The color map to use. Should be a matplotlib colormap object or a string with the name of a matplotlib colormap (default: “magma_r”).</td>
</tr>
<tr class="odd">
<td>vmin</td>
<td>float</td>
<td>0</td>
<td>The value to obtain the minimum color in the colormap. Should be &lt;= np.min(x) to avoid truncation (default: 0).</td>
</tr>
<tr class="even">
<td>vmax</td>
<td>typing.Optional[float]</td>
<td>None</td>
<td>The value to obtain the maximum color in the colormap. Should be &gt;= np.max(x), and will use np.max(x) if set to <code>None</code> (default: None).</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td></td>
<td><strong>Array of RGBA values with a shape of <code>x.shape + (4, )</code>.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L34" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="map_array_to_color" class="level3">
<h3 class="anchored" data-anchor-id="map_array_to_color">map_array_to_color</h3>
<blockquote class="blockquote">
<pre><code> map_array_to_color (x:numpy.ndarray, palette:matplotlib.colors.Colormap,
                     xmax:Optional[float]=None)</code></pre>
</blockquote>
<p>Map an array of values to a color palette.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>x</td>
<td>ndarray</td>
<td></td>
<td>The array to map.</td>
</tr>
<tr class="even">
<td>palette</td>
<td>Colormap</td>
<td></td>
<td>The color map to use. Should be a matplotlib colormap object.</td>
</tr>
<tr class="odd">
<td>xmax</td>
<td>typing.Optional[float]</td>
<td>None</td>
<td>The maximum value to use for normalization. Should be &gt;= np.max(x), and will use np.max(x) if set to <code>None</code> (default: None).</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td></td>
<td><strong>Array of RGBA values with a shape of <code>x.shape + (4, )</code>.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L1006" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_homology_context" class="level3">
<h3 class="anchored" data-anchor-id="add_homology_context">add_homology_context</h3>
<blockquote class="blockquote">
<pre><code> add_homology_context (connections:numpy.ndarray,
                       orthology:pandas.core.frame.DataFrame)</code></pre>
</blockquote>
<p>Add homology context to the given connections based on the orthology information.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>connections</td>
<td>ndarray</td>
<td>The connections between genes. The columns should be (query_gene, target_gene).</td>
</tr>
<tr class="even">
<td>orthology</td>
<td>DataFrame</td>
<td>The orthology information as a DataFrame.</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td><strong>The connections array with homology context added. The columns will be (query_gene,<br>target_gene, connection_strength), and the values in connection_strength will depend on the<br>content of the orthology DataFrame.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L833" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="plot_dotplot" class="level3">
<h3 class="anchored" data-anchor-id="plot_dotplot">plot_dotplot</h3>
<blockquote class="blockquote">
<pre><code> plot_dotplot (query_avg_expr:numpy.ndarray,
               target_avg_expr:numpy.ndarray,
               query_perc_expr:numpy.ndarray,
               target_perc_expr:numpy.ndarray, query_genes:List[str],
               target_genes:List[str], connections:numpy.ndarray,
               query_cluster_colors:Dict[str,str],
               target_cluster_colors:Dict[str,str],
               query_gene_colors:Dict[str,str],
               target_gene_colors:Dict[str,str], query_species:str,
               target_species:str, x_offset:float=1, y_offset:float=0,
               grid_offset:int=30, query_clustering:str='leiden',
               target_clustering:str='leiden',
               output:str='./paired_dotplot.png',
               title:Optional[str]=None, title_font_size:int=16,
               center:bool=True,
               cmap:matplotlib.colors.Colormap='magma_r')</code></pre>
</blockquote>
<p>Plot the paired dotplot based on the given data.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>query_avg_expr</td>
<td>ndarray</td>
<td></td>
<td>Array representing the average expression values of query genes.</td>
</tr>
<tr class="even">
<td>target_avg_expr</td>
<td>ndarray</td>
<td></td>
<td>Array representing the average expression values of target genes.</td>
</tr>
<tr class="odd">
<td>query_perc_expr</td>
<td>ndarray</td>
<td></td>
<td>Array representing the percentage expression values of query genes.</td>
</tr>
<tr class="even">
<td>target_perc_expr</td>
<td>ndarray</td>
<td></td>
<td>Array representing the percentage expression values of target genes.</td>
</tr>
<tr class="odd">
<td>query_genes</td>
<td>typing.List[str]</td>
<td></td>
<td>List of query gene names.</td>
</tr>
<tr class="even">
<td>target_genes</td>
<td>typing.List[str]</td>
<td></td>
<td>List of target gene names.</td>
</tr>
<tr class="odd">
<td>connections</td>
<td>ndarray</td>
<td></td>
<td>An array where each row contains two genes and (optionally) the strength of their<br>connection.</td>
</tr>
<tr class="even">
<td>query_cluster_colors</td>
<td>typing.Dict[str, str]</td>
<td></td>
<td>Dictionary mapping query cluster names to their colors.</td>
</tr>
<tr class="odd">
<td>target_cluster_colors</td>
<td>typing.Dict[str, str]</td>
<td></td>
<td>Dictionary mapping target cluster names to their colors.</td>
</tr>
<tr class="even">
<td>query_gene_colors</td>
<td>typing.Dict[str, str]</td>
<td></td>
<td>Dictionary mapping query gene names to their colors.</td>
</tr>
<tr class="odd">
<td>target_gene_colors</td>
<td>typing.Dict[str, str]</td>
<td></td>
<td>Dictionary mapping target gene names to their colors.</td>
</tr>
<tr class="even">
<td>query_species</td>
<td>str</td>
<td></td>
<td>Species name of the query genes.</td>
</tr>
<tr class="odd">
<td>target_species</td>
<td>str</td>
<td></td>
<td>Species name of the target genes.</td>
</tr>
<tr class="even">
<td>x_offset</td>
<td>float</td>
<td>1</td>
<td>Offset for the x-axis (default: 1).</td>
</tr>
<tr class="odd">
<td>y_offset</td>
<td>float</td>
<td>0</td>
<td>Offset for the y-axis (default: 0).</td>
</tr>
<tr class="even">
<td>grid_offset</td>
<td>int</td>
<td>30</td>
<td>Offset for the grid spacing (default: 30).</td>
</tr>
<tr class="odd">
<td>query_clustering</td>
<td>str</td>
<td>leiden</td>
<td>Clustering method for the query genes (default: “leiden”).</td>
</tr>
<tr class="even">
<td>target_clustering</td>
<td>str</td>
<td>leiden</td>
<td>Clustering method for the target genes (default: “leiden”).</td>
</tr>
<tr class="odd">
<td>output</td>
<td>str</td>
<td>./paired_dotplot.png</td>
<td>Output file path for the plot (default: “./paired_dotplot.png”).</td>
</tr>
<tr class="even">
<td>title</td>
<td>typing.Optional[str]</td>
<td>None</td>
<td>Title of the plot (default: None).</td>
</tr>
<tr class="odd">
<td>title_font_size</td>
<td>int</td>
<td>16</td>
<td>Font size of the plot title (default: 16).</td>
</tr>
<tr class="even">
<td>center</td>
<td>bool</td>
<td>True</td>
<td>Whether to center the dotplots when the number of genes exceeds the maximum (default: True).</td>
</tr>
<tr class="odd">
<td>cmap</td>
<td>Colormap</td>
<td>magma_r</td>
<td></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L760" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="add_connections" class="level3">
<h3 class="anchored" data-anchor-id="add_connections">add_connections</h3>
<blockquote class="blockquote">
<pre><code> add_connections (fig:matplotlib.figure.Figure, connections:numpy.ndarray,
                  query_gene_names:List[str],
                  query_gene_colors:Dict[str,str], label_offset:float)</code></pre>
</blockquote>
<p>Add connections between genes to the given paired dotplot figure.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>fig</td>
<td>Figure</td>
<td>The paired dotplot figure to which the connections will be added.</td>
</tr>
<tr class="even">
<td>connections</td>
<td>ndarray</td>
<td>An array where each row contains two genes and (optionally) the strength of their<br>connection.</td>
</tr>
<tr class="odd">
<td>query_gene_names</td>
<td>typing.List[str]</td>
<td>The list of query gene names.</td>
</tr>
<tr class="even">
<td>query_gene_colors</td>
<td>typing.Dict[str, str]</td>
<td>The dictionary mapping query gene names to their colors.</td>
</tr>
<tr class="odd">
<td>label_offset</td>
<td>float</td>
<td>The offset for label positioning.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L701" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="make_dotplot" class="level3">
<h3 class="anchored" data-anchor-id="make_dotplot">make_dotplot</h3>
<blockquote class="blockquote">
<pre><code> make_dotplot (ax:matplotlib.axes._axes.Axes, avg:numpy.ndarray,
               perc:numpy.ndarray, gene_names:List[str], species:str,
               clustering:str, clust_color:List[str],
               gene_color:List[str], side:str='left',
               cmap:matplotlib.colors.Colormap='magma_r')</code></pre>
</blockquote>
<p>Make a dotplot on the given Axes object based on the average and percentage expression values.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ax</td>
<td>Axes</td>
<td></td>
<td>The Axes object on which to create the dotplot.</td>
</tr>
<tr class="even">
<td>avg</td>
<td>ndarray</td>
<td></td>
<td>The average expression values.</td>
</tr>
<tr class="odd">
<td>perc</td>
<td>ndarray</td>
<td></td>
<td>The percentage expression values.</td>
</tr>
<tr class="even">
<td>gene_names</td>
<td>typing.List[str]</td>
<td></td>
<td>The list of gene names.</td>
</tr>
<tr class="odd">
<td>species</td>
<td>str</td>
<td></td>
<td>The species name.</td>
</tr>
<tr class="even">
<td>clustering</td>
<td>str</td>
<td></td>
<td>The clustering information.</td>
</tr>
<tr class="odd">
<td>clust_color</td>
<td>typing.List[str]</td>
<td></td>
<td>The list of colors for clusters.</td>
</tr>
<tr class="even">
<td>gene_color</td>
<td>typing.List[str]</td>
<td></td>
<td>The list of colors for genes.</td>
</tr>
<tr class="odd">
<td>side</td>
<td>str</td>
<td>left</td>
<td>The side to place the y-axis labels, either “left” or “right” (default: “left”).</td>
</tr>
<tr class="even">
<td>cmap</td>
<td>Colormap</td>
<td>magma_r</td>
<td></td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L672" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="plot_colorbar_legend" class="level3">
<h3 class="anchored" data-anchor-id="plot_colorbar_legend">plot_colorbar_legend</h3>
<blockquote class="blockquote">
<pre><code> plot_colorbar_legend (cbar_legend:matplotlib.axes._axes.Axes,
                       query_avg_expr:numpy.ndarray,
                       target_avg_expr:numpy.ndarray,
                       cmap:matplotlib.colors.Colormap='magma_r')</code></pre>
</blockquote>
<p>Plot the colorbar legend based on the average expression values of query and target genes.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>cbar_legend</td>
<td>Axes</td>
<td></td>
<td>The Axes object representing the colorbar legend.</td>
</tr>
<tr class="even">
<td>query_avg_expr</td>
<td>ndarray</td>
<td></td>
<td>Array representing the average expression values of query genes.</td>
</tr>
<tr class="odd">
<td>target_avg_expr</td>
<td>ndarray</td>
<td></td>
<td>Array representing the average expression values of target genes.</td>
</tr>
<tr class="even">
<td>cmap</td>
<td>Colormap</td>
<td>magma_r</td>
<td>The Colormap instance or registered colormap name used to map scalar data to colors<br>(default: “magma_r”).</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L608" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="plot_dot_legend" class="level3">
<h3 class="anchored" data-anchor-id="plot_dot_legend">plot_dot_legend</h3>
<blockquote class="blockquote">
<pre><code> plot_dot_legend (dot_legend, size_exponent=1.5, dot_size=200)</code></pre>
</blockquote>
<p>Create the dotplot legend, explaining dot size.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dot_legend</td>
<td>matplotlib <code>axis</code></td>
<td></td>
<td>The subplot of the grid that contains the dotplot legend.</td>
</tr>
<tr class="even">
<td>size_exponent</td>
<td>float</td>
<td>1.5</td>
<td>The exponent to raise the fraction of cells in a group to, to get the dot size. The default<br>is 1.5.</td>
</tr>
<tr class="odd">
<td>dot_size</td>
<td>int</td>
<td>200</td>
<td>The size of the largest dot. The default is 200.</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L545" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_dot_color" class="level3">
<h3 class="anchored" data-anchor-id="get_dot_color">get_dot_color</h3>
<blockquote class="blockquote">
<pre><code> get_dot_color (query:anndata._core.anndata.AnnData,
                target:anndata._core.anndata.AnnData,
                query_clustering:str, target_clustering:str,
                query_genes:Optional[numpy.ndarray]=None,
                target_genes:Optional[numpy.ndarray]=None,
                query_gene_names:Optional[numpy.ndarray]=None,
                target_gene_names:Optional[numpy.ndarray]=None,
                layer:Optional[str]=None)</code></pre>
</blockquote>
<p>Calculate average expression in each cluster and translate that to dot color for the dotplot. Note that this function does not know what you did with the matrix before; if you have log-transformed the data it will calculate an average of logs, not the log of the exp-transformed average.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>query</td>
<td>AnnData</td>
<td></td>
<td>The query dataset.</td>
</tr>
<tr class="even">
<td>target</td>
<td>AnnData</td>
<td></td>
<td>The target dataset.</td>
</tr>
<tr class="odd">
<td>query_clustering</td>
<td>str</td>
<td></td>
<td>The .obs column name to use for the query dataset.</td>
</tr>
<tr class="even">
<td>target_clustering</td>
<td>str</td>
<td></td>
<td>The .obs column name to use for the target dataset.</td>
</tr>
<tr class="odd">
<td>query_genes</td>
<td>typing.Optional[numpy.ndarray]</td>
<td>None</td>
<td>Array of query genes to subset the data, if any. If None, use all genes (default: None).</td>
</tr>
<tr class="even">
<td>target_genes</td>
<td>typing.Optional[numpy.ndarray]</td>
<td>None</td>
<td>Array of target genes to subset the data, if any. If None, use all genes (default: None).</td>
</tr>
<tr class="odd">
<td>query_gene_names</td>
<td>typing.Optional[numpy.ndarray]</td>
<td>None</td>
<td>Array of query gene names (default: None).</td>
</tr>
<tr class="even">
<td>target_gene_names</td>
<td>typing.Optional[numpy.ndarray]</td>
<td>None</td>
<td>Array of target gene names (default: None).</td>
</tr>
<tr class="odd">
<td>layer</td>
<td>typing.Optional[str]</td>
<td>None</td>
<td>The layer to use for the average expression calculation. If not specified, it will use the<br><code>.X</code> slot of the <code>AnnData</code> objects. It is vital to set this correctly to avoid calculating<br>average expression on log1p-transformed data (default: None).</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[pandas.core.frame.DataFrame, pandas.core.frame.DataFrame]</strong></td>
<td></td>
<td><strong>A tuple containing the dot color values for query and target datasets, respectively.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L478" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_dot_size" class="level3">
<h3 class="anchored" data-anchor-id="get_dot_size">get_dot_size</h3>
<blockquote class="blockquote">
<pre><code> get_dot_size (query:pandas.core.frame.DataFrame,
               target:pandas.core.frame.DataFrame, query_clustering:str,
               target_clustering:str,
               query_genes:Optional[numpy.ndarray]=None,
               target_genes:Optional[numpy.ndarray]=None,
               query_gene_names:Optional[numpy.ndarray]=None,
               target_gene_names:Optional[numpy.ndarray]=None)</code></pre>
</blockquote>
<p>Calculate which percentage of cells in each cluster express each gene, and translate that to dot size for the dotplot.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>query</td>
<td>DataFrame</td>
<td></td>
<td>The query dataset.</td>
</tr>
<tr class="even">
<td>target</td>
<td>DataFrame</td>
<td></td>
<td>The target dataset.</td>
</tr>
<tr class="odd">
<td>query_clustering</td>
<td>str</td>
<td></td>
<td>The .obs column name to use for the query dataset.</td>
</tr>
<tr class="even">
<td>target_clustering</td>
<td>str</td>
<td></td>
<td>The .obs column name to use for the target dataset.</td>
</tr>
<tr class="odd">
<td>query_genes</td>
<td>typing.Optional[numpy.ndarray]</td>
<td>None</td>
<td>Array of query genes to subset the data, if any. If None, use all genes (default: None).</td>
</tr>
<tr class="even">
<td>target_genes</td>
<td>typing.Optional[numpy.ndarray]</td>
<td>None</td>
<td>Array of target genes to subset the data, if any. If None, use all genes (default: None).</td>
</tr>
<tr class="odd">
<td>query_gene_names</td>
<td>typing.Optional[numpy.ndarray]</td>
<td>None</td>
<td>Array of query gene names (default: None).</td>
</tr>
<tr class="even">
<td>target_gene_names</td>
<td>typing.Optional[numpy.ndarray]</td>
<td>None</td>
<td>Array of target gene names (default: None).</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[pandas.core.frame.DataFrame, pandas.core.frame.DataFrame]</strong></td>
<td></td>
<td><strong>A tuple containing the dot size values for query and target datasets, respectively.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L437" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="feature_colors" class="level3">
<h3 class="anchored" data-anchor-id="feature_colors">feature_colors</h3>
<blockquote class="blockquote">
<pre><code> feature_colors (components:numpy.ndarray, query_G:int, seed:int=42)</code></pre>
</blockquote>
<p>Assign colors to the components based on the given array of components.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>components</td>
<td>ndarray</td>
<td></td>
<td>The array of components.</td>
</tr>
<tr class="even">
<td>query_G</td>
<td>int</td>
<td></td>
<td>The number of components for the query genes.</td>
</tr>
<tr class="odd">
<td>seed</td>
<td>int</td>
<td>42</td>
<td>The seed value for the random number generator (default: 42).</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[numpy.ndarray, numpy.ndarray]</strong></td>
<td></td>
<td><strong>A tuple containing the colored components for query genes and target genes, respectively.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L396" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="gene_order" class="level3">
<h3 class="anchored" data-anchor-id="gene_order">gene_order</h3>
<blockquote class="blockquote">
<pre><code> gene_order (full_adjacency:numpy.ndarray, components:numpy.ndarray,
             query_G:int)</code></pre>
</blockquote>
<p>Calculate the order of genes based on the given full adjacency matrix and components. Highly connected genes are placed first, genes without any connections are randomly ordered in the bottom of the plot.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>full_adjacency</td>
<td>ndarray</td>
<td>The full adjacency matrix represented as a 2D numpy array.</td>
</tr>
<tr class="even">
<td>components</td>
<td>ndarray</td>
<td>An array representing the components.</td>
</tr>
<tr class="odd">
<td>query_G</td>
<td>int</td>
<td>The number of query genes.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[numpy.ndarray, numpy.ndarray]</strong></td>
<td><strong>A tuple containing the query gene order and the target gene order as numpy arrays.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L350" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="calculate_adjacency_matrix" class="level3">
<h3 class="anchored" data-anchor-id="calculate_adjacency_matrix">calculate_adjacency_matrix</h3>
<blockquote class="blockquote">
<pre><code> calculate_adjacency_matrix (connections:numpy.ndarray,
                             query_genes:List[Any],
                             target_genes:List[Any])</code></pre>
</blockquote>
<p>Calculate the adjacency matrix based on the given connections, query genes, and target genes.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>connections</td>
<td>ndarray</td>
<td>The 2D array representing the connections between genes. Each row contains two gene<br>identifiers indicating a connection, and optionally the strength of that connection.</td>
</tr>
<tr class="even">
<td>query_genes</td>
<td>typing.List[typing.Any]</td>
<td>A list of genes that act as queries.</td>
</tr>
<tr class="odd">
<td>target_genes</td>
<td>typing.List[typing.Any]</td>
<td>A list of genes that act as targets.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td><strong>The adjacency matrix represented as a 2D numpy array. It has dimensions (query_G + target_G)<br>x (query_G + target_G), where query_G and target_G are the lengths of query_genes and<br>target_genes, respectively.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L318" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="label_pos" class="level3">
<h3 class="anchored" data-anchor-id="label_pos">label_pos</h3>
<blockquote class="blockquote">
<pre><code> label_pos (display_coords:Dict[str,Tuple[float,float,float,float]],
            key:str, side:str='left')</code></pre>
</blockquote>
<p>Get the edge coordinates of a label. Keep either the left or the right end of the word.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>display_coords</td>
<td>typing.Dict[str, typing.Tuple[float, float, float, float]]</td>
<td></td>
<td>A dictionary that holds the window extents of tick labels.</td>
</tr>
<tr class="even">
<td>key</td>
<td>str</td>
<td></td>
<td>The label to retrieve; a gene name.</td>
</tr>
<tr class="odd">
<td>side</td>
<td>str</td>
<td>left</td>
<td>One of “left” or “right”; depending on orientation will return the leftmost or rightmost<br>position of the label (default: “left”).</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[float, float]</strong></td>
<td></td>
<td><strong>A tuple containing the x and y coordinates of the label.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/galicae/comandos/blob/main/comandos/dotplot_util.py#L269" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="prepare_dotplot" class="level3">
<h3 class="anchored" data-anchor-id="prepare_dotplot">prepare_dotplot</h3>
<blockquote class="blockquote">
<pre><code> prepare_dotplot (avg_expr:pandas.core.frame.DataFrame,
                  perc_expr:pandas.core.frame.DataFrame,
                  cmap:Union[str,matplotlib.colors.Colormap]='magma_r',
                  vmin:float=0, vmax:Optional[float]=None,
                  size_exponent:float=1.5, dot_size:float=200)</code></pre>
</blockquote>
<p>Pivots average expression and percent expressed tables to make them dotplot-friendly.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>avg_expr</td>
<td>DataFrame</td>
<td></td>
<td>Data frame that holds average expression for all genes and all clusters.</td>
</tr>
<tr class="even">
<td>perc_expr</td>
<td>DataFrame</td>
<td></td>
<td>Data frame that tracks the percentage of cells expressing each gene in every cluster.</td>
</tr>
<tr class="odd">
<td>cmap</td>
<td>typing.Union[str, matplotlib.colors.Colormap]</td>
<td>magma_r</td>
<td>The Colormap instance or registered colormap name used to map scalar data to colors<br>(default: “magma_r”).</td>
</tr>
<tr class="even">
<td>vmin</td>
<td>float</td>
<td>0</td>
<td>Minimum average expression value to show (default: 0).</td>
</tr>
<tr class="odd">
<td>vmax</td>
<td>typing.Optional[float]</td>
<td>None</td>
<td>Maximum average expression value to show (default: maximum average expr. value).</td>
</tr>
<tr class="even">
<td>size_exponent</td>
<td>float</td>
<td>1.5</td>
<td>Dot size is computed as fraction ** size_exponent * dot_size (default: 1.5).</td>
</tr>
<tr class="odd">
<td>dot_size</td>
<td>float</td>
<td>200</td>
<td>The size of the largest dot (default: 200).</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[pandas.core.frame.DataFrame, pandas.core.frame.DataFrame, numpy.ndarray]</strong></td>
<td></td>
<td><strong>A tuple containing the melted average expression data frame, the melted percentage<br>expression data frame, and the array of RGBA-coded color values for the average expression<br>in a cluster/gene combination, according to the input color map.</strong></td>
</tr>
</tbody>
</table>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/galicae\.github\.io\/comandos");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/galicae/comandos/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>