{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b08f643",
   "metadata": {},
   "source": [
    "# genes\n",
    "\n",
    "> Prepare and annotate genes and gene sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7fea26a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | default_exp genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "889028a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29243583",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "import os\n",
    "from typing import Union\n",
    "\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa2d6c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "if \"EXAMPLE_DATA_PATH\" not in os.environ.keys():\n",
    "    os.environ[\n",
    "        \"EXAMPLE_DATA_PATH\"\n",
    "    ] = \"/Users/npapadop/Documents/repos/comandos/example_data/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef03b182",
   "metadata": {},
   "source": [
    "SAMap is built on pairwise gene similarity. For cross-species comparisons it is often very\n",
    "interesting to know the level of conservation of two genes - mostly, whether they are orthologs or\n",
    "paralogs. I will provide functions to annotate genes with orthology information from EggNOG-mapper,\n",
    "but you can also use your own orthology annotation, provided it returns the same thing: a $G_1\n",
    "\\times G_2$ table where $G_1$ and $G_2$ are the genes in the two species you are comparing, and each\n",
    "cell $(g_1, g_2)$ contains the orthology relationship between $g_1$ and $g_2$: 2, if they are\n",
    "orthologs, 1 if they are paralogs/in the same gene family, 0 if they are unrelated.\n",
    "\n",
    "This is an extremely inefficient way of saving this matrix, and it would be relatively easy to code\n",
    "this as a sparse matrix, but I don't think it's worth the effort. The matrices are not too large\n",
    "(especially after the gene filtering that is so common to scRNA-seq analysis), and even home\n",
    "computers commonly pack 16GB of RAM these days. Unless you are working with Frankenstein'd genomes\n",
    "with tens of thousands of \"genes\" you will be fine."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfb09dfd",
   "metadata": {},
   "source": [
    "First we will need to read in the EggNOG-mapper result file.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<details>\n",
    "<summary>\n",
    "<b>WARNING - EggNOG format:</b>\n",
    "</summary>\n",
    "\n",
    "Depending on the version of EggNOG you may get a slightly different file; you will need to filter it\n",
    "down to the two columns we need: the query gene ID and the orthogroup assignments. These are a\n",
    "comma-separated string in the format `orthogroup_ID@taxonomic_level`. The taxonomic level will\n",
    "determine whether genes are orthologs or paralogs, so choose wisely. I am using `\"Eukaryota\"` and\n",
    "`\"Bilateria\"` as defaults, but it may well be that your version of EggNOG is using NCBI taxonomic\n",
    "IDs instead of verbose names. Please check before applying!\n",
    "\n",
    "</details>\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<details>\n",
    "<summary>\n",
    "<b>WARNING - Index matching:</b>\n",
    "</summary>\n",
    "\n",
    "For the entries of the EggNOG table to match to the gene names in the SAMap object we need to make\n",
    "sure the index of `query` matches to the index of `sm.sams[query_species].adata.var`. If you created\n",
    "the files in the scheme that I am following, this means that you prepended the species ID to the\n",
    "gene IDs; we would need to do the same here.\n",
    "\n",
    "</details>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f98fef",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/Users/npapadop/Documents/repos/comandos/example_data/eggnog/hydra.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[5]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m query = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m      2\u001b[39m \u001b[43m    \u001b[49m\u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43menviron\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mEXAMPLE_DATA_PATH\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[43m+\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43meggnog/hydra.tsv\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      3\u001b[39m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      4\u001b[39m \u001b[43m    \u001b[49m\u001b[43mengine\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mpython\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      5\u001b[39m \u001b[43m)\u001b[49m\n\u001b[32m      6\u001b[39m query = query[\n\u001b[32m      7\u001b[39m     [\u001b[33m\"\u001b[39m\u001b[33mUnnamed: 0\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33meggNOG_OGs\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m      8\u001b[39m ].copy()  \u001b[38;5;66;03m# I am only keeping the columns I need\u001b[39;00m\n\u001b[32m      9\u001b[39m query.columns = [\u001b[33m\"\u001b[39m\u001b[33mgene_id\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33meggNOG_OGs\u001b[39m\u001b[33m\"\u001b[39m]  \u001b[38;5;66;03m# rename so that it is easier to work with\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Caskroom/miniforge/base/envs/comandos_dev/lib/python3.14/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1013\u001b[39m kwds_defaults = _refine_defaults_read(\n\u001b[32m   1014\u001b[39m     dialect,\n\u001b[32m   1015\u001b[39m     delimiter,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1022\u001b[39m     dtype_backend=dtype_backend,\n\u001b[32m   1023\u001b[39m )\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Caskroom/miniforge/base/envs/comandos_dev/lib/python3.14/site-packages/pandas/io/parsers/readers.py:620\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    617\u001b[39m _validate_names(kwds.get(\u001b[33m\"\u001b[39m\u001b[33mnames\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[32m    619\u001b[39m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m620\u001b[39m parser = \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    622\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[32m    623\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Caskroom/miniforge/base/envs/comandos_dev/lib/python3.14/site-packages/pandas/io/parsers/readers.py:1620\u001b[39m, in \u001b[36mTextFileReader.__init__\u001b[39m\u001b[34m(self, f, engine, **kwds)\u001b[39m\n\u001b[32m   1617\u001b[39m     \u001b[38;5;28mself\u001b[39m.options[\u001b[33m\"\u001b[39m\u001b[33mhas_index_names\u001b[39m\u001b[33m\"\u001b[39m] = kwds[\u001b[33m\"\u001b[39m\u001b[33mhas_index_names\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m   1619\u001b[39m \u001b[38;5;28mself\u001b[39m.handles: IOHandles | \u001b[38;5;28;01mNone\u001b[39;00m = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1620\u001b[39m \u001b[38;5;28mself\u001b[39m._engine = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Caskroom/miniforge/base/envs/comandos_dev/lib/python3.14/site-packages/pandas/io/parsers/readers.py:1880\u001b[39m, in \u001b[36mTextFileReader._make_engine\u001b[39m\u001b[34m(self, f, engine)\u001b[39m\n\u001b[32m   1878\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[32m   1879\u001b[39m         mode += \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m-> \u001b[39m\u001b[32m1880\u001b[39m \u001b[38;5;28mself\u001b[39m.handles = \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1881\u001b[39m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1882\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1883\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1884\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mcompression\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1885\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mmemory_map\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1886\u001b[39m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m=\u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1887\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding_errors\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstrict\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1888\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstorage_options\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1889\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1890\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m.handles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m   1891\u001b[39m f = \u001b[38;5;28mself\u001b[39m.handles.handle\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Caskroom/miniforge/base/envs/comandos_dev/lib/python3.14/site-packages/pandas/io/common.py:873\u001b[39m, in \u001b[36mget_handle\u001b[39m\u001b[34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[39m\n\u001b[32m    868\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m    869\u001b[39m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[32m    870\u001b[39m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[32m    871\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m ioargs.encoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs.mode:\n\u001b[32m    872\u001b[39m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m         handle = \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[32m    874\u001b[39m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    875\u001b[39m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    876\u001b[39m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    877\u001b[39m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    878\u001b[39m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m    879\u001b[39m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    880\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    881\u001b[39m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[32m    882\u001b[39m         handle = \u001b[38;5;28mopen\u001b[39m(handle, ioargs.mode)\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: '/Users/npapadop/Documents/repos/comandos/example_data/eggnog/hydra.tsv'"
     ]
    }
   ],
   "source": [
    "query = pd.read_csv(\n",
    "    os.environ[\"EXAMPLE_DATA_PATH\"] + \"eggnog/hydra.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    engine=\"python\",\n",
    ")\n",
    "query = query[\n",
    "    [\"Unnamed: 0\", \"eggNOG_OGs\"]\n",
    "].copy()  # I am only keeping the columns I need\n",
    "query.columns = [\"gene_id\", \"eggNOG_OGs\"]  # rename so that it is easier to work with\n",
    "query[\"gene_id\"] = \"hy_\" + query[\"gene_id\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f65c91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hy_t33417aep</td>\n",
       "      <td>38ERC@33154,3NUD8@4751,3QR2W@4890,3RR2D@4891,C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hy_t33418aep</td>\n",
       "      <td>2CN11@1,2QT83@2759,38D3X@33154,3BDRF@33208,3CX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hy_t37645aep</td>\n",
       "      <td>38CHP@33154,3B9GY@33208,COG2124@1,KOG0157@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hy_t31628aep</td>\n",
       "      <td>KOG1075@1,KOG1075@2759,KOG4475@1,KOG4475@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hy_t33265aep</td>\n",
       "      <td>3QB9P@4776,COG3145@1,KOG2731@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18265</th>\n",
       "      <td>hy_t24932aep</td>\n",
       "      <td>3A1BV@33154,3BQ1H@33208,KOG1121@1,KOG1121@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18266</th>\n",
       "      <td>hy_t24930aep</td>\n",
       "      <td>2E9XV@1,2SG7Z@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18267</th>\n",
       "      <td>hy_t24940aep</td>\n",
       "      <td>KOG0118@1,KOG0118@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18268</th>\n",
       "      <td>hy_t29557aep</td>\n",
       "      <td>2CZA0@1,2S9AH@2759,3ABW2@33154,3BVK1@33208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18269</th>\n",
       "      <td>hy_t29564aep</td>\n",
       "      <td>COG2340@1,KOG3017@2759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18270 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id                                         eggNOG_OGs\n",
       "0      hy_t33417aep  38ERC@33154,3NUD8@4751,3QR2W@4890,3RR2D@4891,C...\n",
       "1      hy_t33418aep  2CN11@1,2QT83@2759,38D3X@33154,3BDRF@33208,3CX...\n",
       "2      hy_t37645aep     38CHP@33154,3B9GY@33208,COG2124@1,KOG0157@2759\n",
       "3      hy_t31628aep      KOG1075@1,KOG1075@2759,KOG4475@1,KOG4475@2759\n",
       "4      hy_t33265aep                  3QB9P@4776,COG3145@1,KOG2731@2759\n",
       "...             ...                                                ...\n",
       "18265  hy_t24932aep     3A1BV@33154,3BQ1H@33208,KOG1121@1,KOG1121@2759\n",
       "18266  hy_t24930aep                                 2E9XV@1,2SG7Z@2759\n",
       "18267  hy_t24940aep                             KOG0118@1,KOG0118@2759\n",
       "18268  hy_t29557aep         2CZA0@1,2S9AH@2759,3ABW2@33154,3BVK1@33208\n",
       "18269  hy_t29564aep                             COG2340@1,KOG3017@2759\n",
       "\n",
       "[18270 rows x 2 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e857f39e",
   "metadata": {},
   "source": [
    "Now we will filter the `EggNOG_OGs` column and only keep the two levels that we're interested in.\n",
    "Since we are comparing _Hydra_ to a planarian we should be using the Metazoa level for orthologs. We\n",
    "can see from the table visualization that this table uses NCBI tax IDs, so we should look up the tax\n",
    "ID for Metazoa (33208). Similarly, if we would like to use the Eukaryota level for paralogs, we need\n",
    "its tax ID (2759).\n",
    "\n",
    "We also need a function that will filter the orthology table to only keep the OGs that belong to the\n",
    "specified levels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04144078",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def filter_OGs(\n",
    "    x: Union[\n",
    "        list, str\n",
    "    ],  # list of the EggNOG OGs or comma-separated string of EggNOG OGs.\n",
    "    paralog: str = \"Eukaryota\",  # the level of the paralog OG\n",
    "    ortholog: str = \"Bilateria\",  # the level of the ortholog OG\n",
    ") -> list:  # the paralog OG and ortholog OG\n",
    "    \"Find the EggNOG OGs at the the paralog and ortholog level.\"\n",
    "    paralog_OG = \"\"\n",
    "    ortholog_OG = \"\"\n",
    "    if isinstance(x, str):\n",
    "        x = x.split(\",\")\n",
    "    for s in x:\n",
    "        if paralog in s:\n",
    "            paralog_OG = s\n",
    "        if ortholog in s:\n",
    "            ortholog_OG = s\n",
    "    return [paralog_OG, ortholog_OG]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf0d23b8",
   "metadata": {},
   "source": [
    "This function will filter one EggNOG string (or list) to the specified levels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb9e92c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'38ERC@33154,3NUD8@4751,3QR2W@4890,3RR2D@4891,COG5050@1,KOG2877@2759'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query[\"eggNOG_OGs\"].loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0271080",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_str = query[\"eggNOG_OGs\"].loc[0]\n",
    "paralog_str, ortholog_str = filter_OGs(input_str, paralog=\"2759\", ortholog=\"33208\")\n",
    "assert paralog_str == \"KOG2877@2759\"\n",
    "assert ortholog_str == \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "458e5fbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_list = query[\"eggNOG_OGs\"].loc[0].split(\",\")\n",
    "paralog_list, ortholog_list = filter_OGs(input_list, paralog=\"2759\", ortholog=\"33208\")\n",
    "assert paralog_list == \"KOG2877@2759\"\n",
    "assert ortholog_list == \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2aa2243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def assign_homology(\n",
    "    species_OGs,  # the dataframe with the gene_id and the EggNOG OGs\n",
    "    paralog: str = \"Eukaryota\",  # the level of the paralog OG\n",
    "    ortholog: str = \"Bilateria\",  # the level of the ortholog OG)\n",
    ") -> pd.DataFrame:  # the dataframe with the gene_id, paralog OG and ortholog OG\n",
    "    \"Get the taxonomy of the genes.\"\n",
    "    homologs = (\n",
    "        species_OGs[\"eggNOG_OGs\"]\n",
    "        .apply(filter_OGs, paralog=paralog, ortholog=ortholog)\n",
    "        .to_list()\n",
    "    )\n",
    "    taxonomy = pd.DataFrame(homologs, columns=[\"paralog\", \"ortholog\"])\n",
    "    species_OGs = species_OGs.join(taxonomy)\n",
    "\n",
    "    orthologs = pd.DataFrame(\n",
    "        species_OGs.groupby(\"gene_id\", observed=False)[\"ortholog\"].apply(np.unique).str[0]\n",
    "    )\n",
    "    paralogs = pd.DataFrame(\n",
    "        species_OGs.groupby(\"gene_id\", observed=False)[\"paralog\"].apply(np.unique).str[0]\n",
    "    )\n",
    "\n",
    "    result = orthologs.join(paralogs)\n",
    "    result.replace(to_replace=\"\", value=None, inplace=True)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7651793c",
   "metadata": {},
   "outputs": [],
   "source": [
    "hydra_genes = assign_homology(query, paralog=\"2759\", ortholog=\"33208\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c866d0e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ortholog</th>\n",
       "      <th>paralog</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hy_t10003aep</th>\n",
       "      <td>3BCY5@33208</td>\n",
       "      <td>KOG3599@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t10008aep</th>\n",
       "      <td>None</td>\n",
       "      <td>KOG1075@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t10009aep</th>\n",
       "      <td>3BA48@33208</td>\n",
       "      <td>KOG1545@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t10011aep</th>\n",
       "      <td>3BFGW@33208</td>\n",
       "      <td>KOG1136@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t10012aep</th>\n",
       "      <td>3BKRE@33208</td>\n",
       "      <td>KOG2527@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t998aep</th>\n",
       "      <td>3C06S@33208</td>\n",
       "      <td>2S418@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t9990aep</th>\n",
       "      <td>None</td>\n",
       "      <td>KOG0490@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t9992aep</th>\n",
       "      <td>3B9JS@33208</td>\n",
       "      <td>KOG0573@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t999aep</th>\n",
       "      <td>None</td>\n",
       "      <td>2RZ1Z@2759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hy_t99aep</th>\n",
       "      <td>3BH2P@33208</td>\n",
       "      <td>KOG2861@2759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18270 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ortholog       paralog\n",
       "gene_id                                \n",
       "hy_t10003aep  3BCY5@33208  KOG3599@2759\n",
       "hy_t10008aep         None  KOG1075@2759\n",
       "hy_t10009aep  3BA48@33208  KOG1545@2759\n",
       "hy_t10011aep  3BFGW@33208  KOG1136@2759\n",
       "hy_t10012aep  3BKRE@33208  KOG2527@2759\n",
       "...                   ...           ...\n",
       "hy_t998aep    3C06S@33208    2S418@2759\n",
       "hy_t9990aep          None  KOG0490@2759\n",
       "hy_t9992aep   3B9JS@33208  KOG0573@2759\n",
       "hy_t999aep           None    2RZ1Z@2759\n",
       "hy_t99aep     3BH2P@33208  KOG2861@2759\n",
       "\n",
       "[18270 rows x 2 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hydra_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6c5d986",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert hydra_genes.at[\"hy_t10003aep\", \"paralog\"] == \"KOG3599@2759\"\n",
    "assert hydra_genes.at[\"hy_t10003aep\", \"ortholog\"] == \"3BCY5@33208\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39526879",
   "metadata": {},
   "source": [
    "Repeat for the target species (planarian):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abad86c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "target = pd.read_csv(\n",
    "    os.environ[\"EXAMPLE_DATA_PATH\"] + \"eggnog/planarian.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    engine=\"python\",\n",
    ")\n",
    "target = target[\n",
    "    [\"Unnamed: 0\", \"eggNOG_OGs\"]\n",
    "].copy()  # I am only keeping the columns I need\n",
    "target.columns = [\"gene_id\", \"eggNOG_OGs\"]  # rename so that it is easier to work with\n",
    "target[\"gene_id\"] = \"pl_\" + target[\"gene_id\"].astype(str)\n",
    "\n",
    "planarian_genes = assign_homology(target, paralog=\"2759\", ortholog=\"33208\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5040da5e",
   "metadata": {},
   "source": [
    "Given the orthology assignments (orthogroup membership), it is now very easy to calculate which\n",
    "cross-species genes are orthologs or paralogs; we just need to compare the columns of the orthology\n",
    "tables and keep score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a23f919f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | exporti\n",
    "\n",
    "\n",
    "def compare_orthology(query, target):\n",
    "    result = []\n",
    "    for x in query:\n",
    "        result.append(target == x)\n",
    "\n",
    "    result = pd.DataFrame(result)\n",
    "    result.index = query.index\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9789f9f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def calculate_orthology_score(\n",
    "    query: pd.DataFrame,  # the dataframe with the gene_id, paralog OG and ortholog OG for the query species\n",
    "    target: pd.DataFrame,\n",
    ") -> pd.DataFrame:\n",
    "    orthologs = compare_orthology(query[\"ortholog\"], target[\"ortholog\"])\n",
    "    paralogs = compare_orthology(query[\"paralog\"], target[\"paralog\"])\n",
    "    orthology_score = orthologs.astype(int) + paralogs.astype(int)\n",
    "    return orthology_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85d43964",
   "metadata": {},
   "outputs": [],
   "source": [
    "orthology_score = calculate_orthology_score(hydra_genes, planarian_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2d678fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38GXK@33154,3BHNW@33208,3CZKV@33213,47ZE5@7711,48WIG@7742,4CF02@8459,KOG2397@1,KOG2397@2759\n",
      "38GXK@33154,3BHNW@33208,KOG2397@1,KOG2397@2759\n"
     ]
    }
   ],
   "source": [
    "gene1 = \"pl_dd_Smed_v4_10002_0_1\"\n",
    "gene2 = \"hy_t25984aep\"\n",
    "\n",
    "annot1 = target.set_index(\"gene_id\").loc[gene1]\n",
    "annot2 = query.set_index(\"gene_id\").loc[gene2]\n",
    "\n",
    "print(annot1[\"eggNOG_OGs\"])\n",
    "print(annot2[\"eggNOG_OGs\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "578854f1",
   "metadata": {},
   "source": [
    "These genes are orthologs, we should therefore expect them to have an orthology score of 2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e67f56c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert orthology_score[gene1].loc[gene2] == 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3218139c",
   "metadata": {},
   "source": [
    "We can save the orthology table to disk for later use:\n",
    "\n",
    "```python\n",
    "# not run\n",
    "orthology_score.to_csv(\"path/to/hypl_orthology.tsv\", sep=\"\\t\")\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39751c8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def get_orthologs(\n",
    "    genes: np.ndarray, orthology: pd.DataFrame, target: ad.AnnData, celltype_to: str\n",
    ") -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Get orthologous and paralogous gene connections based on the given genes and orthology\n",
    "    information.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    genes : np.ndarray\n",
    "        Array of gene names.\n",
    "    orthology : pd.DataFrame\n",
    "        Data frame representing the orthology information. The index should contain the query genes,\n",
    "        the columns should overlap with the index of target.var, and the values should be 1 for\n",
    "        paralogs and 2 for orthologs.\n",
    "    target : ad.AnnData\n",
    "        Target annotation data.\n",
    "    celltype_to : str\n",
    "        The target cell type. Must be a key in `target.uns[\"rank_genes_groups\"][\"names\"]`.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    np.ndarray\n",
    "        Array of connections between genes, including orthologous and paralogous connections.\n",
    "        Columns are (query, target, degree), where degree is 1 for paralogs and 2 for orthologs.\n",
    "    \"\"\"\n",
    "    genes_in_table = np.intersect1d(genes, orthology.index)\n",
    "\n",
    "    to_include = np.sum(orthology.loc[genes_in_table]) == 2\n",
    "    orthologs = np.intersect1d(orthology.columns[to_include], target.var.index)\n",
    "\n",
    "    to_include = np.sum(orthology.loc[genes_in_table]) == 1\n",
    "    paralogs = np.intersect1d(orthology.columns[to_include], target.var.index)\n",
    "\n",
    "    scores = pd.DataFrame(target.uns[\"rank_genes_groups\"][\"scores\"])[celltype_to]\n",
    "    names = pd.DataFrame(target.uns[\"rank_genes_groups\"][\"names\"])[celltype_to]\n",
    "\n",
    "    framed = pd.DataFrame(scores).set_index(names)\n",
    "    significant = framed.loc[paralogs] > 0.01\n",
    "    homologs = np.concatenate((orthologs, paralogs[significant[celltype_to]]))\n",
    "\n",
    "    subset = orthology.loc[genes_in_table][homologs].melt(ignore_index=False)\n",
    "    subset.reset_index(drop=False, inplace=True)\n",
    "\n",
    "    connections = np.array(subset[subset[\"value\"] > 0])\n",
    "\n",
    "    not_in_table = np.setdiff1d(genes, orthology.index)\n",
    "    if not_in_table.size > 0:\n",
    "        unconnected = np.array([[g, None, 0] for g in not_in_table])\n",
    "        return np.concatenate((connections, unconnected))\n",
    "    else:\n",
    "        return connections\n",
    "\n",
    "\n",
    "def get_orthologs_overlap(genes1, genes2, query, target, orthology):\n",
    "    \"\"\"\n",
    "    Returns a DataFrame of homologous gene pairs between two sets of genes based on their presence\n",
    "    in an orthology table.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    genes1 : numpy.ndarray\n",
    "        A series of gene names.\n",
    "    genes2 : numpy.ndarray\n",
    "        A series of gene names.\n",
    "    query : anndata.AnnData\n",
    "        An AnnData object containing the query genes as indices of the `.var` slot.\n",
    "    target : anndata.AnnData\n",
    "        An AnnData object containing the target genes as indices of the `.var` slot.\n",
    "    orthology : pandas.core.frame.DataFrame\n",
    "        A DataFrame containing the orthology information.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    connections : pandas.core.frame.DataFrame\n",
    "        A DataFrame of homologous gene pairs and their degree of conservation. The array has\n",
    "        three columns: 'query', 'target', and 'degree', where 'query' and 'target' are the gene\n",
    "        names, and 'degree' is the degree of conservation, which can be either 1 or 2.\n",
    "    \"\"\"\n",
    "    genes1_in_data = np.intersect1d(genes1, query.var.index)\n",
    "    genes2_in_data = np.intersect1d(genes2, target.var.index)\n",
    "    genes1_in_table = np.intersect1d(genes1_in_data, orthology.index)\n",
    "    genes2_in_table = np.intersect1d(genes2_in_data, orthology.columns)\n",
    "\n",
    "    connections = []\n",
    "\n",
    "    subset = orthology.loc[genes1_in_table][genes2_in_table]\n",
    "    _qo, _to = np.where(subset == 2)\n",
    "    for q, t in zip(subset.index[_qo], subset.columns[_to]):\n",
    "        connections.append([q, t, \"2\"])\n",
    "\n",
    "    _qo, _to = np.where(subset == 1)\n",
    "    for q, t in zip(subset.index[_qo], subset.columns[_to]):\n",
    "        connections.append([q, t, \"1\"])\n",
    "\n",
    "    return np.array(connections)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dd870cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "import nbdev\n",
    "\n",
    "nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
