{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dotplot_util\n",
    "\n",
    "> Collection of utility functions for the pairwise dotplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | default_exp dotplot_util"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fastcore.docments import docments\n",
    "from fastcore.test import *\n",
    "# | hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib import cm\n",
    "from matplotlib.cm import ScalarMappable\n",
    "from matplotlib.colors import Normalize\n",
    "from matplotlib.patches import ConnectionPatch\n",
    "from numpy.random import default_rng\n",
    "from scipy.sparse.csgraph import connected_components\n",
    "\n",
    "try:\n",
    "    from matplotlib.colors import TwoSlopeNorm as DivNorm\n",
    "except ImportError:\n",
    "    # matplotlib<3.2\n",
    "    from matplotlib.colors import DivergingNorm as DivNorm\n",
    "\n",
    "from comandos import util"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General plotting utility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | exporti\n",
    "\n",
    "\n",
    "def map_array_to_color(x, palette, xmax=None):\n",
    "    if xmax is None:\n",
    "        xmax = np.max(x)\n",
    "    color = palette(x / xmax)\n",
    "    return color\n",
    "\n",
    "\n",
    "def map_to_colormap(x, cmap=\"magma_r\", vmin=0, vmax=None):\n",
    "    if vmax is None:\n",
    "        vmax = np.max(x)\n",
    "    norm = mpl.colors.Normalize(vmin=vmin, vmax=vmax)\n",
    "    m = cm.ScalarMappable(norm=norm, cmap=cmap)\n",
    "    return m.to_rgba(x)\n",
    "\n",
    "\n",
    "def unique_genes(connections):\n",
    "    real = connections != np.array(None)\n",
    "    return np.unique(connections[real])\n",
    "\n",
    "\n",
    "def check_colornorm(vmin=None, vmax=None, vcenter=None, norm=None):\n",
    "    # from ScanPy\n",
    "    if norm is not None:\n",
    "        if (vmin is not None) or (vmax is not None) or (vcenter is not None):\n",
    "            raise ValueError(\"Passing both norm and vmin/vmax/vcenter is not allowed.\")\n",
    "    else:\n",
    "        if vcenter is not None:\n",
    "            norm = DivNorm(vmin=vmin, vmax=vmax, vcenter=vcenter)\n",
    "        else:\n",
    "            norm = Normalize(vmin=vmin, vmax=vmax)\n",
    "\n",
    "    return norm\n",
    "\n",
    "\n",
    "def highlight_cluster(clusters, cluster, bg=\"black\", hl=\"red\"):\n",
    "    clust_col = [bg] * len(clusters)\n",
    "    if cluster is None or cluster == \"\":\n",
    "        print(\"Invalid cluster name; returning background color.\")\n",
    "        return clust_col\n",
    "    if cluster not in clusters:\n",
    "        print(\"Cluster not found; returning background color.\")\n",
    "        return clust_col\n",
    "    highlight = np.where(clusters == cluster)[0][0]\n",
    "    clust_col[highlight] = hl\n",
    "    return clust_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "godsnot_102 = np.array(\n",
    "    [\n",
    "        \"#FFFF00\",\n",
    "        \"#1CE6FF\",\n",
    "        \"#FF34FF\",\n",
    "        \"#FF4A46\",\n",
    "        \"#008941\",\n",
    "        \"#006FA6\",\n",
    "        \"#A30059\",\n",
    "        \"#FFDBE5\",\n",
    "        \"#7A4900\",\n",
    "        \"#0000A6\",\n",
    "        \"#63FFAC\",\n",
    "        \"#B79762\",\n",
    "        \"#004D43\",\n",
    "        \"#8FB0FF\",\n",
    "        \"#997D87\",\n",
    "        \"#5A0007\",\n",
    "        \"#809693\",\n",
    "        \"#6A3A4C\",\n",
    "        \"#1B4400\",\n",
    "        \"#4FC601\",\n",
    "        \"#3B5DFF\",\n",
    "        \"#4A3B53\",\n",
    "        \"#FF2F80\",\n",
    "        \"#61615A\",\n",
    "        \"#BA0900\",\n",
    "        \"#6B7900\",\n",
    "        \"#00C2A0\",\n",
    "        \"#FFAA92\",\n",
    "        \"#FF90C9\",\n",
    "        \"#B903AA\",\n",
    "        \"#D16100\",\n",
    "        \"#DDEFFF\",\n",
    "        \"#000035\",\n",
    "        \"#7B4F4B\",\n",
    "        \"#A1C299\",\n",
    "        \"#300018\",\n",
    "        \"#0AA6D8\",\n",
    "        \"#013349\",\n",
    "        \"#00846F\",\n",
    "        \"#372101\",\n",
    "        \"#FFB500\",\n",
    "        \"#C2FFED\",\n",
    "        \"#A079BF\",\n",
    "        \"#CC0744\",\n",
    "        \"#C0B9B2\",\n",
    "        \"#C2FF99\",\n",
    "        \"#001E09\",\n",
    "        \"#00489C\",\n",
    "        \"#6F0062\",\n",
    "        \"#0CBD66\",\n",
    "        \"#EEC3FF\",\n",
    "        \"#456D75\",\n",
    "        \"#B77B68\",\n",
    "        \"#7A87A1\",\n",
    "        \"#788D66\",\n",
    "        \"#885578\",\n",
    "        \"#FAD09F\",\n",
    "        \"#FF8A9A\",\n",
    "        \"#D157A0\",\n",
    "        \"#BEC459\",\n",
    "        \"#456648\",\n",
    "        \"#0086ED\",\n",
    "        \"#886F4C\",\n",
    "        \"#34362D\",\n",
    "        \"#B4A8BD\",\n",
    "        \"#00A6AA\",\n",
    "        \"#452C2C\",\n",
    "        \"#636375\",\n",
    "        \"#A3C8C9\",\n",
    "        \"#FF913F\",\n",
    "        \"#938A81\",\n",
    "        \"#575329\",\n",
    "        \"#00FECF\",\n",
    "        \"#B05B6F\",\n",
    "        \"#8CD0FF\",\n",
    "        \"#3B9700\",\n",
    "        \"#04F757\",\n",
    "        \"#C8A1A1\",\n",
    "        \"#1E6E00\",\n",
    "        \"#7900D7\",\n",
    "        \"#A77500\",\n",
    "        \"#6367A9\",\n",
    "        \"#A05837\",\n",
    "        \"#6B002C\",\n",
    "        \"#772600\",\n",
    "        \"#D790FF\",\n",
    "        \"#9B9700\",\n",
    "        \"#549E79\",\n",
    "        \"#FFF69F\",\n",
    "        \"#201625\",\n",
    "        \"#72418F\",\n",
    "        \"#BC23FF\",\n",
    "        \"#99ADC0\",\n",
    "        \"#3A2465\",\n",
    "        \"#922329\",\n",
    "        \"#5B4534\",\n",
    "        \"#FDE8DC\",\n",
    "        \"#404E55\",\n",
    "        \"#0089A3\",\n",
    "        \"#CB7E98\",\n",
    "        \"#A4E804\",\n",
    "        \"#324E72\",\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def prepare_dotplot(\n",
    "    avg_expr,\n",
    "    perc_expr,\n",
    "    cmap=\"magma_r\",\n",
    "    vmin=0,\n",
    "    vmax=None,\n",
    "    size_exponent=1.5,\n",
    "    dot_size=200,\n",
    "):\n",
    "    \"\"\"Pivots average expression and percent expressed tables to make them dotplot-friendly.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    avg_expr : pandas DataFrame\n",
    "        Data frame that holds average expression for all genes and all clusters.\n",
    "    perc_expr : pandas DataFrame\n",
    "        Data frame that tracks the percentage of cells expressing each gene in every cluster.\n",
    "    cmap : str or Colormap, optional\n",
    "        The Colormap instance or registered colormap name used to map scalar data to colors, by\n",
    "        default \"magma_r\"\n",
    "    vmin : int, optional\n",
    "        Minimum average expression value to show, by default 0\n",
    "    vmax : _type_, optional\n",
    "        Maximum average expression value to show, by default None (defaults to maximum average expr.\n",
    "        value).\n",
    "    size_exponent : float, optional\n",
    "        Dot size is computed as fraction ** size_exponent * dot_size. By default 1.5.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    df_avg_expr : pandas DataFrame\n",
    "        Melted data frame. Each row contains a cluster name, gene name, and the average expression\n",
    "        value of that gene in that cluster.\n",
    "    df_perc_expr : pandas DataFrame\n",
    "        Melted data frame. Each row contains a cluster name, gene name, and the percentage of cells\n",
    "        expressing the gene in that cluster.\n",
    "    color : numpy array\n",
    "        An array that holds RGBA-coded color values for the average expression in a cluster/gene\n",
    "        combination, according to the input color map.\n",
    "    \"\"\"\n",
    "    df_avg_expr = avg_expr.reset_index().melt(\"index\")\n",
    "    df_avg_expr.columns = [\"row\", \"column\", \"value\"]\n",
    "    color = map_to_colormap(df_avg_expr[\"value\"], cmap=cmap, vmin=vmin, vmax=vmax)\n",
    "\n",
    "    df_perc_expr = perc_expr.reset_index().melt(\"index\")\n",
    "    df_perc_expr.columns = [\"row\", \"column\", \"value\"]\n",
    "    df_perc_expr[\"value\"] = df_perc_expr[\"value\"] ** size_exponent * dot_size\n",
    "    return df_avg_expr, df_perc_expr, color\n",
    "\n",
    "\n",
    "def label_pos(display_coords, key, side=\"left\"):\n",
    "    \"\"\"Get the edge coordinates of a label. Keep either the left or the right end of the word.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    display_coords : dict\n",
    "        A dictionary that holds the window extents of tick labels.\n",
    "    key : str\n",
    "        The label to retrieve; a gene name.\n",
    "    side : str, optional\n",
    "        One of \"left\" or \"right\"; depending on orientation will return the leftmost or rightmost\n",
    "        position of the label. By default \"left\"\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    x : float\n",
    "        The left or right edge of the label.\n",
    "    y : float\n",
    "        The middle of the label.\n",
    "    \"\"\"\n",
    "    x0, y0, x1, y1 = display_coords[key].flatten()\n",
    "    if side == \"left\":\n",
    "        x = x1\n",
    "    else:\n",
    "        x = x0\n",
    "    y = (y0 + y1) / 2\n",
    "    return x, y\n",
    "\n",
    "\n",
    "def calculate_adjacency_matrix(connections, query_genes, target_genes):\n",
    "    query_G = len(query_genes)\n",
    "    target_G = len(target_genes)\n",
    "    adjacency = np.zeros(\n",
    "        (query_G, target_G), dtype=bool\n",
    "    )  # not a true adjacency matrix; targetely the top right corner.\n",
    "    query_connected = {gene: i for i, gene in enumerate(query_genes)}\n",
    "    target_connected = {gene: i for i, gene in enumerate(target_genes)}\n",
    "\n",
    "    if connections.shape[1] == 3:\n",
    "        connections = connections[:, :2]\n",
    "    for [g1, g2] in connections:\n",
    "        if g1 is None or g2 is None:\n",
    "            continue\n",
    "        i1 = query_connected[g1]\n",
    "        i2 = target_connected[g2]\n",
    "        adjacency[i1, i2] = True\n",
    "\n",
    "    full_adjacency = np.zeros((query_G + target_G, query_G + target_G), dtype=bool)\n",
    "    full_adjacency[:query_G][:, query_G:] = adjacency\n",
    "    full_adjacency[query_G:][:, :query_G] = adjacency.T\n",
    "    return full_adjacency\n",
    "\n",
    "\n",
    "def gene_order(full_adjacency, components, query_G):\n",
    "    comp, freq = np.unique(components, return_counts=True)\n",
    "    degrees = np.sum(full_adjacency, axis=0)\n",
    "\n",
    "    keep = freq > 1\n",
    "    descending = np.argsort(-freq[freq > 1])\n",
    "\n",
    "    no_comps = len(comp[freq > 1])\n",
    "    for i, c in enumerate(comp[keep][descending]):\n",
    "        # degrees[components == c] *= freq[keep][descending][i] * (no_comps - i)\n",
    "        degrees[components == c] = no_comps - i\n",
    "\n",
    "    query_degree = degrees[:query_G]\n",
    "    target_degree = degrees[query_G:]\n",
    "\n",
    "    query_order = np.argsort(-query_degree)\n",
    "    target_order = np.argsort(-target_degree)\n",
    "    return query_order, target_order\n",
    "\n",
    "\n",
    "def feature_colors(components, query_G, seed=42):\n",
    "    comp, freq = np.unique(components, return_counts=True)\n",
    "    rng = default_rng(seed)\n",
    "    no_comps = len(comp[freq > 1])\n",
    "    if no_comps < 102:\n",
    "        pick_colors = rng.choice(len(godsnot_102), size=no_comps, replace=False)\n",
    "    else:\n",
    "        pick_colors = rng.choice(len(godsnot_102), size=no_comps, replace=True)\n",
    "    godsnot_keep = {}\n",
    "    for i, component in enumerate(comp[freq > 1]):\n",
    "        godsnot_keep[component] = godsnot_102[pick_colors][i]\n",
    "\n",
    "    components_colored = [\"black\"] * len(components)\n",
    "    for i, component in enumerate(components):\n",
    "        if component in comp[freq > 1]:\n",
    "            components_colored[i] = godsnot_keep[component]\n",
    "\n",
    "    query_comp_color = np.array(components_colored[:query_G])\n",
    "    target_comp_color = np.array(components_colored[query_G:])\n",
    "    return query_comp_color, target_comp_color\n",
    "\n",
    "\n",
    "def get_dot_size(\n",
    "    query,\n",
    "    target,\n",
    "    query_clustering,\n",
    "    target_clustering,\n",
    "    query_genes=None,\n",
    "    target_genes=None,\n",
    "    query_gene_names=None,\n",
    "    target_gene_names=None,\n",
    "):\n",
    "    query_cluster_size = query.obs[query_clustering].value_counts()\n",
    "    target_cluster_size = target.obs[target_clustering].value_counts()\n",
    "\n",
    "    query_perc_expr = (\n",
    "        util.grouped_obs_present(query, query_clustering) / query_cluster_size\n",
    "    )\n",
    "    target_perc_expr = (\n",
    "        util.grouped_obs_present(target, target_clustering) / target_cluster_size\n",
    "    )\n",
    "\n",
    "    # subset to only the genes we have\n",
    "    if query_genes is not None and target_genes is not None:\n",
    "        query_perc_expr = query_perc_expr.loc[query_genes].copy().T\n",
    "        target_perc_expr = target_perc_expr.loc[target_genes].copy().T\n",
    "    else:\n",
    "        print(\"No genes supplied; returning all genes.\")\n",
    "\n",
    "    if query_gene_names is not None:\n",
    "        query_perc_expr.columns = query.var[query_gene_names].loc[\n",
    "            query_perc_expr.columns\n",
    "        ]\n",
    "    if target_gene_names is not None:\n",
    "        target_perc_expr.columns = target.var[target_gene_names].loc[\n",
    "            target_perc_expr.columns\n",
    "        ]\n",
    "    return query_perc_expr, target_perc_expr\n",
    "\n",
    "\n",
    "def get_dot_color(\n",
    "    query,\n",
    "    target,\n",
    "    query_clustering,\n",
    "    target_clustering,\n",
    "    query_genes=None,\n",
    "    target_genes=None,\n",
    "    query_gene_names=None,\n",
    "    target_gene_names=None,\n",
    "):\n",
    "    query_avg_expr = util.grouped_obs_mean(query, query_clustering)\n",
    "    target_avg_expr = util.grouped_obs_mean(target, target_clustering)\n",
    "    # subset to only the genes we have\n",
    "    if query_genes is not None and target_genes is not None:\n",
    "        query_avg_expr = query_avg_expr.loc[query_genes].copy().T\n",
    "        target_avg_expr = target_avg_expr.loc[target_genes].copy().T\n",
    "    else:\n",
    "        print(\"No genes supplied; returning all genes.\")\n",
    "\n",
    "    if query_gene_names is not None:\n",
    "        query_avg_expr.columns = query.var[query_gene_names].loc[query_avg_expr.columns]\n",
    "    if target_gene_names is not None:\n",
    "        target_avg_expr.columns = target.var[target_gene_names].loc[\n",
    "            target_avg_expr.columns\n",
    "        ]\n",
    "    return query_avg_expr, target_avg_expr\n",
    "\n",
    "\n",
    "def plot_dot_legend(\n",
    "    dot_legend,\n",
    "    size_exponent=1.5,\n",
    "    dot_size=200,\n",
    "):\n",
    "    \"\"\"Create the dotplot legend, explaining dot size.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dot_legend : matplotlib `axis`\n",
    "        The subplot of the grid that contains the dotplot legend.\n",
    "    size_exponent : float, optional\n",
    "        The exponent to raise the fraction of cells in a group to, to get the dot size. The default\n",
    "        is 1.5.\n",
    "    dot_size : int, optional\n",
    "        The size of the largest dot. The default is 200.\n",
    "    \"\"\"\n",
    "    dot_legend.axis(\"equal\")\n",
    "    xticks = np.arange(5)\n",
    "    dot_legend.scatter(\n",
    "        xticks,\n",
    "        np.zeros(5),\n",
    "        s=np.arange(0.2, 1.1, 0.2) ** size_exponent * dot_size,\n",
    "        c=\"gray\",\n",
    "        edgecolors=\"black\",\n",
    "    )\n",
    "    dot_legend.set_xlim(-0.5, 4.5)\n",
    "    dot_legend.yaxis.set_visible(False)\n",
    "    dot_legend.spines[[\"left\", \"right\", \"top\", \"bottom\"]].set_visible(False)\n",
    "    dot_legend.xaxis.set_ticks(xticks)\n",
    "    dot_legend.xaxis.set_ticklabels([\"20\", \"40\", \"60\", \"80\", \"100\"])\n",
    "    dot_legend.set_title(\"Fraction of cells\\nin group (%)\", fontsize=\"small\")\n",
    "\n",
    "\n",
    "def _get_mappable(query_avg_expr, target_avg_expr, cmap=\"magma_r\"):\n",
    "    all_expr = np.concatenate(\n",
    "        (query_avg_expr.values.flatten(), target_avg_expr.values.flatten())\n",
    "    )\n",
    "    max_expr = np.max(all_expr)\n",
    "    norm = check_colornorm(0, max_expr)\n",
    "    return ScalarMappable(norm, cmap=cmap)\n",
    "\n",
    "\n",
    "def plot_colorbar_legend(cbar_legend, query_avg_expr, target_avg_expr, cmap=\"magma_r\"):\n",
    "    mappable = _get_mappable(query_avg_expr, target_avg_expr, cmap=cmap)\n",
    "    mpl.colorbar.Colorbar(cbar_legend, mappable=mappable, orientation=\"horizontal\")\n",
    "    cbar_legend.yaxis.set_visible(False)\n",
    "    cbar_legend.set_title(\"Mean expression\\nin group\", fontsize=\"small\")\n",
    "    cbar_legend.xaxis.set_tick_params(labelsize=\"small\")\n",
    "\n",
    "\n",
    "def make_dotplot(\n",
    "    ax,\n",
    "    avg,\n",
    "    perc,\n",
    "    gene_names,\n",
    "    species,\n",
    "    clustering,\n",
    "    clust_color,\n",
    "    gene_color,\n",
    "    side=\"left\",\n",
    "):\n",
    "    N, G = avg.shape\n",
    "    df_avg, df_perc, color = prepare_dotplot(avg, perc)\n",
    "    ax.scatter(\n",
    "        df_avg[\"row\"],\n",
    "        df_avg[\"column\"],\n",
    "        c=color,\n",
    "        s=df_perc[\"value\"],\n",
    "        edgecolors=\"black\",\n",
    "    )\n",
    "    ax.axis(\"equal\")\n",
    "    ax.set_title(f\"species: {species}, clustering: {clustering}\")\n",
    "    ax.xaxis.set_ticks(np.arange(N))\n",
    "    ax.xaxis.set_ticklabels(avg.index, rotation=90)\n",
    "    ax.yaxis.set_ticks(np.arange(G), labels=gene_names)\n",
    "    if side == \"left\":\n",
    "        ax.yaxis.tick_right()\n",
    "    else:\n",
    "        ax.yaxis.tick_left()\n",
    "    ax.invert_yaxis()\n",
    "    [t.set_color(clust_color[i]) for i, t in enumerate(ax.xaxis.get_ticklabels())]\n",
    "    [t.set_color(gene_color[i]) for i, t in enumerate(ax.yaxis.get_ticklabels())]\n",
    "\n",
    "\n",
    "def add_connections(\n",
    "    fig, connections, query_gene_names, query_gene_colors, label_offset\n",
    "):\n",
    "    label_offset = np.array([label_offset, 0])\n",
    "    left = fig.get_axes()[0]\n",
    "    right = fig.get_axes()[1]\n",
    "\n",
    "    left_labels = {}\n",
    "    for label in left.get_yticklabels():\n",
    "        bbox = label.get_window_extent()\n",
    "        coords = fig.transFigure.inverted().transform(bbox)\n",
    "        left_labels[label.get_text()] = coords + label_offset\n",
    "\n",
    "    right_labels = {}\n",
    "    for label in right.get_yticklabels():\n",
    "        bbox = label.get_window_extent()\n",
    "        coords = fig.transFigure.inverted().transform(bbox)\n",
    "        right_labels[label.get_text()] = coords - label_offset\n",
    "\n",
    "    query_index = {g: i for i, g in enumerate(query_gene_names)}\n",
    "    for c in connections:\n",
    "        linestyle = \"dotted\"\n",
    "        if len(c) == 3:\n",
    "            gene_from, gene_to, ls = c\n",
    "            if ls == 0:\n",
    "                linestyle = (0, (1, 5))\n",
    "            elif ls == 1:\n",
    "                linestyle = (0, (5, 5))\n",
    "            elif ls == 2:\n",
    "                linestyle = \"solid\"\n",
    "        else:\n",
    "            gene_from, gene_to = c\n",
    "        if gene_from is None or gene_to is None:\n",
    "            continue\n",
    "        species_a = label_pos(left_labels, gene_from, side=\"left\")\n",
    "        species_b = label_pos(right_labels, gene_to, side=\"right\")\n",
    "        con = ConnectionPatch(\n",
    "            xyA=species_a,\n",
    "            xyB=species_b,\n",
    "            coordsA=\"figure fraction\",\n",
    "            coordsB=\"figure fraction\",\n",
    "            color=query_gene_colors[query_index[gene_from]],\n",
    "            linestyle=linestyle,\n",
    "        )\n",
    "        right.add_artist(con)\n",
    "\n",
    "\n",
    "def _plot_dotplot(\n",
    "    query_avg_expr,\n",
    "    target_avg_expr,\n",
    "    query_perc_expr,\n",
    "    target_perc_expr,\n",
    "    query_genes,\n",
    "    target_genes,\n",
    "    connections,\n",
    "    query_cluster_colors,\n",
    "    target_cluster_colors,\n",
    "    query_gene_colors,\n",
    "    target_gene_colors,\n",
    "    query_species,\n",
    "    target_species,\n",
    "    x_offset=1,\n",
    "    y_offset=0,\n",
    "    grid_offset=30,\n",
    "    query_clustering=\"leiden\",\n",
    "    target_clustering=\"leiden\",\n",
    "    output=\"./paired_dotplot.png\",\n",
    "    title=None,\n",
    "    title_font_size=16,\n",
    "    center=True,\n",
    "):\n",
    "    # make shortcuts for the number of clusters and number of genes\n",
    "    query_N, query_G = query_avg_expr.shape\n",
    "    target_N, target_G = target_avg_expr.shape\n",
    "\n",
    "    # sometimes we end up overcorrecting when batch-producing\n",
    "    # images this way\n",
    "    y_size = np.max([query_G, target_G]) / 2 + y_offset\n",
    "    if y_size < 0:\n",
    "        y_size = np.max([query_G, target_G]) / 2\n",
    "\n",
    "    # create empty figure that scales with number of clusters and genes\n",
    "    figsize = ((query_N + target_N) / 2 + x_offset, y_size)\n",
    "\n",
    "    fig = plt.figure(figsize=figsize)\n",
    "    max_genes = np.max([query_G, target_G])\n",
    "    columns = query_N + target_N + grid_offset + 6\n",
    "    ax = fig.add_gridspec(max_genes, columns)\n",
    "\n",
    "    # add space for the left dotplot (query)\n",
    "    left_start = 0\n",
    "    if center and query_G > max_genes:\n",
    "        left_start = (max_genes - query_G) // 2 - 1\n",
    "    left_end = left_start + query_G\n",
    "    left = fig.add_subplot(ax[left_start:left_end, 0:query_N])\n",
    "\n",
    "    # add space for the right dotplot (target)\n",
    "    right_start = 0\n",
    "    if center and target_G > max_genes:\n",
    "        right_start = (max_genes - target_G) // 2 - 1\n",
    "    right_end = right_start + target_G\n",
    "\n",
    "    after_blank = query_N + grid_offset\n",
    "    before_legends = query_N + grid_offset + target_N\n",
    "    right = fig.add_subplot(\n",
    "        ax[\n",
    "            right_start:right_end,\n",
    "            after_blank:before_legends,\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    left.set_zorder(0)\n",
    "    right.set_zorder(1)\n",
    "\n",
    "    make_dotplot(\n",
    "        left,\n",
    "        query_avg_expr,\n",
    "        query_perc_expr,\n",
    "        query_genes,\n",
    "        query_species,\n",
    "        query_clustering,\n",
    "        query_cluster_colors,\n",
    "        query_gene_colors,\n",
    "        side=\"left\",\n",
    "    )\n",
    "\n",
    "    make_dotplot(\n",
    "        right,\n",
    "        target_avg_expr,\n",
    "        target_perc_expr,\n",
    "        target_genes,\n",
    "        target_species,\n",
    "        target_clustering,\n",
    "        target_cluster_colors,\n",
    "        target_gene_colors,\n",
    "        side=\"right\",\n",
    "    )\n",
    "\n",
    "    # saving the figure: don't forget the dpi option!\n",
    "    if title is not None:\n",
    "        fig.suptitle(title, fontsize=title_font_size)\n",
    "    fig.savefig(output)\n",
    "\n",
    "    # calculate label offset to make the links between genes more legible\n",
    "    label_offset = 1 / (query_N + target_N + grid_offset * 2) / 3\n",
    "    # draw appropriately colored lines between connected genes\n",
    "    add_connections(fig, connections, query_genes, query_gene_colors, label_offset)\n",
    "\n",
    "    dot_start = right_start + target_G // 3\n",
    "    dot_legend = fig.add_subplot(ax[dot_start : (dot_start + 1), -5:])\n",
    "    plot_dot_legend(dot_legend)\n",
    "\n",
    "    cbar_start = right_start + 2 * target_G // 3\n",
    "    cbar_legend = fig.add_subplot(ax[cbar_start : (cbar_start + 1), -3:])\n",
    "    plot_colorbar_legend(cbar_legend, query_avg_expr, target_avg_expr)\n",
    "\n",
    "    plt.savefig(output)\n",
    "\n",
    "\n",
    "def add_homology_context(connections, orthology):\n",
    "    query_genes = [g for g in connections[:, 0] if g is not None]\n",
    "    target_genes = [g for g in connections[:, 1] if g is not None]\n",
    "\n",
    "    query_orthology = np.intersect1d(query_genes, orthology.index)\n",
    "    target_orthology = np.intersect1d(target_genes, orthology.columns)\n",
    "\n",
    "    subset = (\n",
    "        orthology.loc[query_orthology][target_orthology]\n",
    "        .melt(ignore_index=False)\n",
    "        .reset_index(drop=False)\n",
    "    )\n",
    "\n",
    "    homologs = subset[subset[\"value\"] > 0]\n",
    "    homologs.columns = [\"query_id\", \"target_id\", \"value\"]\n",
    "    query_no_homology = np.setdiff1d(query_genes, homologs[\"query_id\"])\n",
    "    target_no_homology = np.setdiff1d(target_genes, homologs[\"target_id\"])\n",
    "\n",
    "    query_side = np.concatenate(\n",
    "        (\n",
    "            homologs[\"query_id\"].values,\n",
    "            query_no_homology,\n",
    "            [None] * len(target_no_homology),\n",
    "        )\n",
    "    )\n",
    "    target_side = np.concatenate(\n",
    "        (\n",
    "            homologs[\"target_id\"].values,\n",
    "            [None] * len(query_no_homology),\n",
    "            target_no_homology,\n",
    "        )\n",
    "    )\n",
    "    orthology_info = np.concatenate(\n",
    "        (\n",
    "            homologs[\"value\"].values,\n",
    "            [0] * (len(query_no_homology) + len(target_no_homology)),\n",
    "        )\n",
    "    )\n",
    "\n",
    "    connections = np.array([query_side, target_side, orthology_info]).T\n",
    "    return connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "import nbdev\n",
    "\n",
    "nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
